pkg/client/rootd/dns/server.go

```diff
@@ -370,15 +373,37 @@ func (d dfs) String() string {
 }
 
 func (s *Server) performRecursionCheck(c context.Context) {
+	const maxRetry = 10
+	defer dlog.Debug(c, "Recursion check finished")
 	rc := strings.TrimSuffix(recursionCheck, ".")
 	dlog.Debugf(c, "Performing initial recursion check with %s", rc)
-	if _, err := net.DefaultResolver.LookupHost(c, rc); err != nil {
-		if err, ok := err.(*net.DNSError); ok {
-			if err.IsNotFound {
-				return
+	i := 0
+	for ; i < maxRetry; i++ {
+		_, err := net.DefaultResolver.LookupIP(c, "ip4", rc)
+		if err != nil {
+			if derr, ok := err.(*net.DNSError); ok && derr.IsNotFound {
+				err = nil
 			}
 		}
-		dlog.Errorf(c, "recursion check ended with %v", err)
+		if err != nil {
+			dlog.Errorf(c, "recursion check ended with %v", err)
+		}
+		if atomic.LoadInt32(&s.recursive) != 99 {
+			return
+		}
+		// Check didn't hit our resolver. Try again after a second
+		dtime.SleepWithContext(c, time.Second)
+
+		// Check that the resolver didn't get hit during our wait. We don't want
+		// to retry if it did, because that will give the false impression that
+		// the resolver is recursive.
+		if atomic.LoadInt32(&s.recursive) != 99 {
+			return
+		}
```

Comment:
i'm like okay with this the way it's implemented because it's realistic but i do feel the need to point out that there's a potential race condition here --say the resolver, for whatever reason, doesn't get hit until exactly when we hit line 403, we'll  end up producing that false impression anyway

i don't think there's a need to really address that though, because a DNS query to localhost should really not take longer than a second